packages/opentelemetry-core/src/utils/environment.ts

```diff
@@ -79,7 +101,7 @@ export function parseEnvironment(values: ENVIRONMENT_MAP): ENVIRONMENT {
         break;
 
       case 'OTEL_LOG_LEVEL':
-        parseNumber(key, environment, values, LogLevel.ERROR, LogLevel.DEBUG);
+        parseLogLevel(key, environment, values);
```

Comment:
Really don't love the side-effect driven nature of this.

Reply:
I don't understand what is meant by that? Could you please elaborate?

Reply:
You run a function which has no output, but modifies global state (the environment). This state then affects the behavior of the program down the line. Instead, I think a better solution is to have a function which looks at the env and returns a loglevel. This function is then simply called during the creation of a logger and the environment is never modified.

Reply:
I'm starting to get it, the `parseNumber()` function used for OTEL_SAMPLING_PROBABILITY (and previously to this PR for OTEL_LOG_LEVEL), is a side-effect driver function or not though?

Reply:
Yes. `parseDriver` modifies global state, which is the definition of side-effect.

Reply:
you mean `parseNumber`

Reply:
So, I based the log level setting function on the `parseNumber` behaviour which is side-effect driven.

Should I do something non side-effect driven for log level in this PR?

Reply:
Yes. I am suggesting something like the following:

```typescript
function getLogLevelFromEnv() {
  // no need to pass key and values in since it will always come from the same var
  const value = environment.OTEL_LOG_LEVEL;
  switch (typeof value === 'string' ? value.toUpperCase() : value) {
    case 'DEBUG':
      return LogLevel.DEBUG;

    case 'INFO':
      return LogLevel.INFO;

    case 'WARN':
      return LogLevel.WARN;

    case 'ERROR':
      return LogLevel.ERROR;

    default:
      return null;
  }
}
```

This function to be used when determining which log level to use when creating a logger.

Reply:
very nice, lastly, this new function should be exported, and we will not use `getEnv()` anymore for logger instantiation, but rather this new function?

Reply:
The whole idea of having the getEnv() is to have one place to get you all env parsed, validated with default values etc. If you use any of this function then you endup of problem for cross platform and repeating the same code again and again in many places. So using any of "internal" function just doesn't make sense with this whole idea.

Reply:
With regards to "side effect" we have exactly the same pattern with injecting / extracting context. This is correct here, too you just want to have one function that will validate the setting and if the setting is valid then add that to the env so that default value doesn't overwrite it later. Not sure why then "side effect" is wrong here ?

Reply:
Please advise @dyladan, @obecny raises a good point

Reply:
Imagine the following case:

1. user sets the env var to "warning" which is invalid
2. user sees info level logs
3. as part of debugging, they print the env var and see "2" because we have overwritten it, which is confusing
4. they spend time trying to figure out why their env var is failing to be set, when the real problem is that the value itself is wrong

If others think this isn't a problem, then I am willing to concede.

Reply:
You are mixing here 2 things, in this function (helper internal function) you are adding a new param to private internal variable not a process.env - nothing modifies that.

Reply:
ah ok

Reply:
https://github.com/open-telemetry/opentelemetry-js/blob/e9b2cf9aeb1daf5ffbab800681bfe1cafc636576/packages/opentelemetry-api/src/context/propagation/setter.ts#L29

Reply:
and here
https://github.com/open-telemetry/opentelemetry-js/blob/e9b2cf9aeb1daf5ffbab800681bfe1cafc636576/packages/opentelemetry-core/src/context/propagation/B3Propagator.ts#L47

Reply:
That's a good example @dyladan, @obecny are you holding your ground on this one? I'll let you make the final call (for now at least) if nobody else wants to weigh in.

Reply:
I have a feeling we are talking about 2 different thing, If I understand correctly, @dyladan was thinking that `parse...` helpers function modifies the `process.env` which is not the case at all. All the helpers are doing they are just adding / injecting values into a variable that you tell it to inject. In this example this is a new variable that takes all config/env variables from either process.env or window on browser and returns a valid values (validated, defaulted etc.) but in both cases it doesn't modify the process.env or window at all. In fact you can provide any obj and tell it to give you the validated params. The whole pattern how the helpers are behaving is the same approach what we use in propagators. You create a new carrier (our internal env variable) and tell it to inject all valid environment params.

Reply:
ok so if I understand correctly, @dyladan was afraid that we were modified `process.env` or browser equivalent window, however we never were, and so there is no danger of `side-effect`? as such current implementation is ok?

Reply:
yes. i was mistaken

Reply:
In fact we are simply modifying an internal env var/carrier called `environment` right?

Reply:
I don't know lets @dyladan comment, I'm calling a day (friday night btw) for me today ;)

Reply:
Have a great weekend @obecny 🕺 ! And thanks for the explanations.